### PR TITLE
REF: eliminate method _write() in json writers

### PR DESCRIPTION
**Pull request by ivanovmg**
- [ ] closes #xxxx
- [ ] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

Attempted to figure out what is wrong with the issue https://github.com/pandas-dev/pandas/issues/36211.
I noticed that the private methods ``_write`` are defined in each child of class ``Writer``, which simply re-define ``obj`` and call the parent ``_write`` method.
It turns out that the source of the recursion does not lie there, but I refactored the module slightly by removing the mediator ``_write`` method. Instead I put the differences between the writers into their ``__init__`` method.

It seems that source of the problem with the recursion mentioned in the bug report is in ``pandas/_libs/src/ujson/python/objToJSON.c``.

### CODE REVIEW

File: pandas/io/json/_json.py
Comment (WillAyd) [inline pandas/io/json/_json.py:188]: Hmm why do we need to do this now?

Comment (ivanovmg) [reply pandas/io/json/_json.py:188]: I do not know. I simply used the original implementation and moved whatever assignments were in ``_write`` into ``__init__``.

Comment (WillAyd) [reply pandas/io/json/_json.py:188]: Oh OK - I see what was done here. Let's not assign to `self.obj` as that will mess with the idempotency

Comment (ivanovmg) [reply pandas/io/json/_json.py:188]: We assign to ``self.obj`` at the moment of the writer initialization, not when calling a method. So, I would argue that it is not harmful.

Comment (WillAyd) [reply pandas/io/json/_json.py:188]: I still think it is confusing to change what `self.obj` is for certain orients - would prefer not to do this

**Comment by ivanovmg** [comment]
If there is a problem with updating ``self.obj`` in ``__init__`` of child classes, then we probably should abandon this PR. Maybe other opinion?

**Comment by WillAyd** [comment]
Yea I don't think this is an improvement as it is better for the subclasses to override the implementation where appropriate rather than modifying the object that gets stored during initialization.

Let's keep open for a few days as someone may disagree

**Comment by ivanovmg** [comment]
Would you consider another implementation? I created a new property ``obj_to_write`` (we can call it a ``table_obj``, or something like this). This allowed us to avoid re-definition of ``self.obj``, reduce ``__init__`` boilerplate and also (as initially) get rid of ``_write`` method.

File: pandas/io/json/_json.py
Comment (WillAyd) [inline pandas/io/json/_json.py:163]: Can you add an annotation for this?

**Review by WillAyd — CHANGES_REQUESTED** [review]
Yea I think this is a better approach

File: pandas/io/json/_json.py
Comment (ivanovmg) [reply pandas/io/json/_json.py]: Added.

**Review by WillAyd — APPROVED** [review]
lgtm

File: pandas/io/json/_json.py
Comment (jbrockmendel) [inline pandas/io/json/_json.py:38]: should this be FrameOrSeries?

File: pandas/io/json/_json.py
Comment (jbrockmendel) [inline pandas/io/json/_json.py:166]: can you explicitly `pass` here

are the keys in the Dict necessarily str, or can they be any Index entries?

**Comment by jbrockmendel** [comment]
nitpick about annotations, otherwise LGTM

File: pandas/io/json/_json.py
Comment (ivanovmg) [reply pandas/io/json/_json.py:38]: The problem with ``FrameOrSeries`` is that it does not allow changing the type of return value.
Inside the function ``to_json`` there is a statement, which changes object from series to frame, which is not allowed.
```
    if orient == "table" and isinstance(obj, Series):
        obj = obj.to_frame(name=obj.name or "values")
```

mypy error:
```
pandas/io/json/_json.py:78: error: Incompatible types in assignment (expression has type "DataFrame", variable has type "FrameOrSeries")  [assignment]
```

I tried to use ``FrameOrSeriesUnion``, which allows changing type from series to frame, but I get another mypy error.
Apparently for ``NDFrame`` only ``FrameOrSeries`` is a correct casting.

```
pandas/core/generic.py:2159: error: The erased type of self "Union[pandas.core.frame.DataFrame, pandas.core.series.Series]" is not a supertype of its class "pandas.core.generic.NDFrame"  [misc]
```

So, I have no clue how to deal with typing here.

File: pandas/io/json/_json.py
Comment (ivanovmg) [reply pandas/io/json/_json.py:166]: Oh yes, actually the keys in the dictionary can be of ``IndexLabel`` type.
In ``FrameWriter`` the resulting dictionary keys can be tuples of strings in case of multicolumns.
In other cases the keys seems to be only strings.

**Comment by ivanovmg** [comment]
All checks are good after merging with master, except Windows:
```
##[error]We stopped hearing from agent Azure Pipelines 11. 
```

@jbrockmendel, unfortunately, I cannot fully resolve the issue with typing. Please see my comment above in relation to ``NDFrame`` and ``FrameOrSeries``.

File: pandas/io/json/_json.py
Comment (jbrockmendel) [reply pandas/io/json/_json.py:43]: @willayd i expected FrameOrSeriesUnion would work for this, am i missing something?

Comment (WillAyd) [reply pandas/io/json/_json.py:43]: @ivanovmg can you merge master and update the error message here? I don't think that line number is reflective of the latest code base

I think the error message is stemming from a mixture of using a Union and a TypeVar (the latter being somewhere in generic). Ideally should be cleaned up but I don't think needs to hold up this PR

Comment (ivanovmg) [reply pandas/io/json/_json.py:43]: Here we go.
I merged master, then modified ``pandas/core/generic.py``.

line 2164:
```
    def to_json(
        self: FrameOrSeriesUnion,
        path_or_buf: Optional[FilePathOrBuffer] = None,
```

And replaced ``NDFrame`` with ``FrameOrSeriesUnion``.

The error would be:
```
pandas/core/generic.py:2164: error: The erased type of self "Union[pandas.core.frame.DataFrame, pandas.core.series.Series]" is not a supertype of its class "pandas.core.generic.NDFrame"  [misc]
```

However, if I do not touch ``generic.py``, then the following error with be thrown.
```
pandas/core/generic.py:2437: error: Argument "obj" to "to_json" has incompatible type "NDFrame"; expected "Union[DataFrame, Series]"  [arg-type]
```

**Comment by WillAyd** [comment]
Thanks @ivanovmg

**Pull request merged**
